packages/grafana-schema/src/common/data_frame.cue

```diff
@@ -0,0 +1,217 @@
+package common
+
+//  See also:
+//  https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/frame_type.go
+DataFrameType:
+	"timeseries-wide" |
+	"timeseries-long" |
+	 // @deprecated in favor of TimeSeriesMulti
+	"timeseries-many" |
+	"timeseries-multi" |
+	// Directory listing
+	"directory-listing" |
+	// First field is X, the rest are ordinal values used as rows in the heatmap
+	"heatmap-rows" |
+	//Explicit fields for xMin, yMin, count, ...
+	//All values in the grid exist and have regular spacing
+	//If the y value is actually ordinal, use `meta.custom` to specify the bucket lookup values
+	"heatmap-cells" |
+	//Explicit fields for xMin, xMax, count
+	"histogram" @cuetsy(kind="enum", memberNames="TimeSeriesWide|TimeSeriesLong|TimeSeriesMany|TimeSeriesMulti|DirectoryListing|HeatmapRows|HeatmapCells|Histogram")
+
+// TODO docs
+DataFrame: {
+	QueryResultBase
+  name?: string
+  // All fields of equal length
+  fields: [...Field]
+  // The number of rows
+  length: int64
+} @cuetsy(kind="interface")
+
+// TODO Field<T = any, V = Vector<T>>
+Field: {
```

Comment:
Not sure about this one. `Vector` is actually an interface wit ha lots of methods, so not really serialisable.

Reply:
the DataFrame interface as is is explicitly not serializable (directly) originally it was apache arrow, but I think we can/should remove the interface, it will require something like https://github.com/grafana/grafana/pull/38615

cc @leeoniya -- that is something we should consider for 10, but as is the `DataFrame` type in both frontend+backend do not correspond to implementations that thema will generate easily.  (Totally possible, but a bigger project). The metadata side is much more approachable.

Reply:
> as is the DataFrame type in both frontend+backend do not correspond to implementations that thema will generate easily

_resists the urge to "CHALLENGE ACCEPTED!"_ 😉 

It sounds like there are specific, known concerns here? AFAIK our only real issue atm is with recursive TS types. Discriminated unions aren't really an issue anymore, given TS veneers